crates/red_knot_python_semantic/resources/mdtest/scopes/moduletype_attrs.md

```diff
@@ -0,0 +1,134 @@
+# Implicit globals from `types.ModuleType`
+
+## Implicit `ModuleType` globals
+
+All modules are instances of `types.ModuleType`.
+If a name can't be found in any local or global scope, we look it up
+as an attribute on `types.ModuleType` in typeshed
+before deciding that the name is unbound.
+
+```py
+reveal_type(__name__)  # revealed: str
+
+# TODO: infer types from PEP-604 union annotations
+reveal_type(__file__)  # revealed: @Todo
+reveal_type(__loader__)  # revealed: @Todo
+reveal_type(__package__)  # revealed: @Todo
+reveal_type(__spec__)  # revealed: @Todo
+
+# TODO: generics
+reveal_type(__path__)  # revealed: @Todo
+
+# TODO: this should probably be added to typeshed; not sure why it isn't?
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__doc__)
+
+class X:
+    reveal_type(__name__)  # revealed: str
+
+def foo():
+    reveal_type(__name__)  # revealed: str
+```
+
+However, three attributes on `types.ModuleType` are not present as implicit
+module globals; these are excluded:
+
+```py path=unbound_dunders.py
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__getattr__)
+
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__dict__)
+
+# error: [unresolved-reference]
+# revealed: Unbound
+reveal_type(__init__)
+```
+
+## Accessed as attributes
+
+`ModuleType` attributes can also be accessed as attributes on module-literal types.
+The special attributes `__dict__` and `__init__`, and all attributes on
+`builtins.object`, can also be accessed as attributes on module-literal types,
+despite the fact that these are inaccessible as globals from inside the module:
+
+```py
+import typing
+
+reveal_type(typing.__name__)  # revealed: str
+reveal_type(typing.__init__)  # revealed: Literal[__init__]
+
+# This one comes from `builtins.object`, not `types.ModuleType`:
+# TODO: we don't currently understand `types.ModuleType` as inheriting from `object`
+reveal_type(typing.__eq__)  # revealed: Unbound
+
+# TODO: needs support for attribute access on instances
+reveal_type(typing.__dict__)  # revealed: @Todo
+```
+
+Typeshed includes a fake `__getattr__` method in the stub for `types.ModuleType`
+to help out with dynamic imports; but we ignore that for module-literal types
+where we know exactly which module we're dealing with:
+
+```py path=__getattr__.py
+import typing
+
+reveal_type(typing.__getattr__)  # revealed: Unbound
+```
+
+## `types.ModuleType.__dict__` takes precedence over global variable `__dict__`
+
+It's impossible to override the `__dict__` attribute of `types.ModuleType`
+instances from inside the module; we should prioritise the attribute in
+the `types.ModuleType` stub over a variable named `__dict__` in the module's
+global namespace:
+
+```py path=foo.py
+__dict__ = "foo"
```

Comment:
can we also test here that accessing `__dict__` within the module, subsequent to this, does get `Literal["foo"]`? But `from foo import __dict__ as d` gives you the module dictionary?

Reply:
> But `from foo import __dict__ as d` gives you the module dictionary?

Woah, TIL. That's nuts.

Reply:
From imports are just syntax sugar for an attribute access!